elephant/signal_processing.py

```diff
@@ -239,6 +239,82 @@ def _design_butterworth_filter(Fs, hpfreq=None, lpfreq=None, order=4):
         return filtered_data
 
 
+def wavelet_transform(signal, freq, nco, fs):
+    """
+    Compute the wavelet transform of a given signal with Morlet mother wavelet.
+    The definition of the wavelet is based on Le van Quyen et al. J
+    Neurosci Meth 111:83-98 (2001).
+
+    Parameters
+    ----------
+    signal : 1D array_like
+        Signal to be wavelet-transformed
+    freq : float
+        Center frequency of the Morlet wavelet.
+    nco : float
+        Size of the mother wavelet (approximate number of cycles within a
+        wavelet). A larger nco value leads to a higher frequency resolution but
+        a lower temporal resolution, and vice versa. Typically used values are
+        in a range of 3 - 8.
+    fs : float
+        Sampling rate of the signal.
+
+    Returns
+    -------
+    signal_trans: 1D complex array_like
+        Wavelet-transformed signal
+    """
+    # Morlet wavelet generator (c.f. Le van Quyen et al. J Neurosci Meth
+    # 111:83-98 (2001))
+    def _morlet_wavelet(freq, nco, Fs, N):
```

Comment:
Wouldn't it be computationally more efficient to determine the Morlet wavelet directly in k-space instead of writing it in time domain and then performing a fft? In the following is an example taken from Torrence & Compo's code, also available on github under https://github.com/chris-torrence/wavelets

```
def wave_bases(mother, k, scale, param):
    n = len(k)
    kplus = np.array(k > 0., dtype=float)

    if mother == 'MORLET':  # -----------------------------------  Morlet

        if param == -1:
            param = 6.

        k0 = np.copy(param)
        expnt = -(scale * k - k0) ** 2 / 2. * kplus
        norm = np.sqrt(scale * k[1]) * (np.pi ** (-0.25)) * \
                np.sqrt(n)  # total energy=N   [Eqn(7)]
        daughter = norm * np.exp(expnt)
        daughter = daughter * kplus  # Heaviside step function
        fourier_factor = (4 * np.pi) / (k0 + np.sqrt(2 + k0 ** 2)
                                            )  # Scale-->Fourier [Sec.3h]
        coi = fourier_factor / np.sqrt(2)  # Cone-of-influence [Sec.3g]
        dofmin = 2  # Degrees of freedom

    return daughter, fourier_factor, coi, dofmin
```

Reply:
they also calculate the cone of influence in a following step according to

```
coi = coi * dt * np.concatenate((np.insert(np.arange((n1 + 1) / 2 - 1), [0], [1E-5]),
				               np.insert(np.flipud(np.arange(0, n1 / 2 - 1)), [-1], [1E-5])))  # COI [Sec.3g]
```
The cone of influence is a very helpful output for the user.